app/controllers/renewal/WhatYouNeedController.scala

```diff
@@ -15,29 +15,41 @@
  */
 
 package controllers.renewal
-
+import cats.data.OptionT
+import cats.implicits._
+import connectors.DataCacheConnector
 import controllers.{AmlsBaseController, CommonPlayDependencies}
 import javax.inject.{Inject, Singleton}
+import models.businessmatching.{BusinessActivities, BusinessMatching, BusinessMatchingMsbServices}
 import models.registrationprogress.{NotStarted, Section, Started}
-import play.api.mvc.MessagesControllerComponents
+import play.api.mvc.{MessagesControllerComponents, Request}
 import services.RenewalService
 import utils.AuthAction
 import views.html.renewal._
 
-import scala.concurrent.ExecutionContext
+import scala.concurrent.{ExecutionContext, Future}
 import scala.concurrent.ExecutionContext.Implicits.global
-
 @Singleton
-class WhatYouNeedController @Inject()(val authAction: AuthAction,
-                                      val ds: CommonPlayDependencies,
-                                      renewalService: RenewalService,
-                                      val cc: MessagesControllerComponents)(implicit ec: ExecutionContext) extends AmlsBaseController(ds, cc) {
-
+class WhatYouNeedController @Inject()(
+                                       val dataCacheConnector: DataCacheConnector,
+                                       val authAction: AuthAction,
+                                       val ds: CommonPlayDependencies,
+                                       renewalService: RenewalService,
+                                       val cc: MessagesControllerComponents)(implicit ec: ExecutionContext) extends AmlsBaseController(ds, cc) {
   def get = authAction.async {
     implicit request =>
-      renewalService.getSection(request.credId) map {
-        case Section(_,NotStarted | Started,_,_) => Ok(what_you_need())
-        case _ => Redirect(controllers.routes.RegistrationProgressController.get())
-      }
+        (for {
+          cache <- OptionT(dataCacheConnector.fetch[BusinessMatching](request.credId, BusinessMatching.key))
+          ba <- OptionT.fromOption[Future](cache.activities)
+          section <- OptionT.liftF(getSection(renewalService, request.credId, Some(ba), cache.msbServices))
+        } yield {
+          section
+        }).getOrElse(Redirect(controllers.routes.RegistrationProgressController.get()))
```

Comment:
As stated above. Do we want to do this? Is this error recoverable? If not we likely want to raise an exception such that we can investigate. If we encounter a non recoverable error such as failing to locate business matching in the cache then sending the user back to the progress page might be confusing? The user will keep trying to get into the section and could get stuck in a loop without an exception being thrown therefore giving us nothing to investigate. 

If we need this data to be able to proceed then should we raise an error if this data can't be located? 

The alternative is to load a default if the data can't be retrieved.

Reply:
Done